src/TextDateOfBirthInput/TextDateOfBirthInput.tsx

```diff
@@ -0,0 +1,140 @@
+import React, { useEffect, useState } from 'react'
+import styled from 'styled-components'
+
+import { TextInput } from '../TextInput'
+import { Box } from '../Box'
+import { Fieldset, FieldsetProps } from '../fields/Fieldset'
+import { dateOfBirthValidator } from '../utils/dateOfBirth/dateOfBirthValidator'
+import { createErrorMessage } from '../utils/dateOfBirth/schema'
+
+export interface DateObjectType {
+  day: string
+  month: string
+  year: string
+}
+
+export type DateObject = {
+  [K in keyof DateObjectType]?: DateObjectType[K] | null
+}
+
+export type TextDateOfBirthInputProps = {
+  value: DateObjectType
+  onChange: (value: DateObjectType) => void
+  showCompleted?: boolean
+} & Pick<FieldsetProps, 'label' | 'error' | 'assistiveText'>
+
+/**
+ * Renders a set of input fields to collect a date of birth as separate day, month,
+ * and year fields.
+ *
+ * @function dateOfBirthValidator validates a date of birth provided in a DateObject format:
+ * - All fields are present and correctly formatted.
+ * - The age is at least 17 years old.
+ * - The year is not before 1900.
+ */
+export const TextDateOfBirthInput = ({
+  onChange,
+  value,
+  label,
+  assistiveText,
+  error,
+  showCompleted = false,
+}: TextDateOfBirthInputProps) => {
+  const [errorMsg, setErrorMsg] = useState<string | undefined>(undefined)
+  const [fieldsWithError, setFieldsWithError] = useState<string[]>([])
+
+  const handleValidationErrors = () => {
+    const validationResult = dateOfBirthValidator(value)
+
+    if ('error' in validationResult) {
+      const message = createErrorMessage(
+        validationResult.error,
+        validationResult.fields,
+      )
+      setErrorMsg(message)
+      setFieldsWithError(validationResult.fields || [])
+    } else {
+      setErrorMsg(undefined)
+      setFieldsWithError([])
+    }
+  }
+
+  useEffect(() => {
+    handleValidationErrors()
+  }, [value])
```

Comment:
Correct me if I am wrong but the component but you are planning to use zod in order to implement this?

If so, I think this logic should sit outside of the component and the component should be stateless, taking the state from react hook form. To clarify, I think `dateOfBirthValidator` should be use within the schema.